pkg/snap/util/calico.go

```diff
@@ -32,7 +32,11 @@ func MaybePatchCalicoAutoDetectionMethod(ctx context.Context, s snap.Snap, canRe
 	}
 
 	var re *regexp.Regexp
-	if ip := net.ParseIP(canReachHost); ip.To4() == nil {
+	ip := net.ParseIP(canReachHost)
+	if ip == nil {
+		return fmt.Errorf("%q is not a valid IP string representation", canReachHost)
+	}
+	if ip.To4() == nil {
```

Comment:
I would recommend a different approach. If the IP is not valid, then we can simply do nothing instead of failing hard on our tracks. Would you prefer failing hard, or handling this with some grace with something like:

```go
ip := net.ParseIP(canReachHost)
switch {
case ip == nil:
   // maybe log something
  return nil
case ip.To4() == nil:
  re = ip6AutodetectionMethodRe
default:
  re = ipAutodetectionMethodRe
}

// continue
```

Reply:
I'm not 100% familiar with the Calico IP autodetection feature. However, according to the methods doc string it seems that having "first-found" enabled for a multi-node cluster seems to cause issues.

If we handle bad IPs gracefully then we would still use the "first-found" method but still joining the cluster, thus creating NIC issues, right?

That was my reason to go for the hard failing but I might be wrong :)